src/mlpack/methods/reinforcement_learning/replay/hindsight_replay.hpp

```diff
@@ -0,0 +1,423 @@
+   */
+  HindsightReplay(const size_t batchSize,
+                    const size_t capacity,
+                    const size_t herRatio = 4,
+                    EnvironmentType environment = EnvironmentType(),
+                    goalStrategy strategy = goalStrategy::FUTURE,
+                    const size_t nSteps = 1,
+                    const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity((herRatio+1)*capacity),
+      position(0),
+      full(false),
+      herRatio(herRatio),
+      goalSelectionStrategy(strategy),
+      nSteps(nSteps),
+      states(dimension, (herRatio+1)*capacity),
+      actions((herRatio+1)*capacity),
+      rewards((herRatio+1)*capacity),
+      nextStates(dimension, (herRatio+1)*capacity),
+      goals(dimension, (herRatio+1)*capacity),
+      isTerminal((herRatio+1)*capacity),
+      environment(environment)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Sample goal according to goal strategy
+   * 
+   * @param transitionIndex index of current transition
+   */
+  void SampleGoal(StateType& desiredGoal, int transitionIndex)
+  {
+    if (goalSelectionStrategy == goalStrategy::FINAL)
+    {
+      // Sample goal as the final state of the epsiode
+      desiredGoal = episodeTransitions[episodeTransitions.size() -1].nextState;
+    }
+    else if (goalSelectionStrategy == goalStrategy::FUTURE)
+    {
+      // Sample goal as random state from that transition to end of epsiode
+      size_t index = arma::randi(
+        arma::distr_param(transitionIndex+1, episodeTransitions.size() -1));
+      desiredGoal = episodeTransitions[index].nextState;
+    }
+    else if (goalSelectionStrategy == goalStrategy::EPISODE)
+    {
+      // Sample goal as random state from the whole epsiode
+      desiredGoal = episodeTransitions[arma::randi(
+        arma::distr_param(0, episodeTransitions.size() -1))].nextState;
+    }
+  }
+
+  /**
+   * Sample goals according to goal strategy and HER ratio
+   * 
+   * @param desiredGoals desired goals as per goal strategy
+   * @param  transitionIndex index of current transition
+   */
+  void SampleGoals(std::vector<StateType>& desiredGoals, int transitionIndex)
+  { 
+    StateType desiredGoal;
+
+    // Store HER transitions based on strategy for each transition in epsiode
+    for (size_t goalIndex = 0; goalIndex < herRatio; ++goalIndex)
+    {
+      SampleGoal(desiredGoal, transitionIndex);
+
+      desiredGoals.push_back(desiredGoal);
+    }
+  }
+
+  /**
+   * Sample goal according to goal strategy
+   * @param discount The discount parameter.
+   */
+  void StoreHERTransitions(const double& discount)
+  { 
+    std::vector<Transition> baseTransitions = episodeTransitions;
+
+    // cannot create a future hindsight goal in the last transition of episode
+    if (goalSelectionStrategy == goalStrategy::FUTURE)
+    {
+      baseTransitions.pop_back();
+    }
+
+    std::vector<StateType> desiredGoals;
```

Comment:
declare inside for loop

Reply:
I think we can do that too. Let me try that out.  There shouldnt be any issue

Reply:
> declare inside for loop, remove `desiredGoals.clear()`, or make `SampleGoals` return `std::vector<StateType>`?
> 
> ```c++
> const std::vector<StateType> desiredGoals = SampleGoals(transitionIndex);
> ```

I actually did it my way so that we avoid having to initialise one more sample goals vector inside the function. And we could directly set a the values into the main `sampleGoals`